src/Storages/MergeTree/PartMetadataCache.cpp

```diff
@@ -0,0 +1,136 @@
+#include "PartMetadataCache.h"
+
+#if USE_ROCKSDB
+#include <rocksdb/db.h>
+#include <Disks/IDisk.h>
+#include <IO/HashingReadBuffer.h>
+#include <IO/ReadHelpers.h>
+#include <IO/ReadBufferFromString.h>
+#include <IO/ReadBufferFromFileBase.h>
+#include <Storages/MergeTree/IMergeTreeDataPart.h>
+#include <Storages/MergeTree/MergeTreeMetadataCache.h>
+
+namespace ProfileEvents
+{
+    extern const Event MergeTreeMetadataCacheHit;
+    extern const Event MergeTreeMetadataCacheMiss;
+}
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+namespace DB
+{
+
+std::unique_ptr<SeekableReadBuffer>
+PartMetadataCache::readOrSet(const DiskPtr & disk, const String & file_name, String & value)
+{
+    String file_path = fs::path(getFullRelativePath()) / file_name;
+    auto status = cache->get(file_path, value);
+    if (!status.ok())
+    {
+        ProfileEvents::increment(ProfileEvents::MergeTreeMetadataCacheMiss);
+        if (!disk->exists(file_path))
```

Comment:
Is it possible by the way?

Reply:
When using cache, it has the possibility that `file_path` may not exists. 

See code: 
``` cpp 

void IMergeTreeDataPart::loadUUID()
{
    String path = fs::path(getFullRelativePath()) / UUID_FILE_NAME;
    bool exists = false;
    std::unique_ptr<SeekableReadBuffer> in;

#if USE_ROCKSDB
    String _;
    if (use_metadata_cache)
    {
        in = metadata_cache->readOrSet(volume->getDisk(), UUID_FILE_NAME, _);
        exists = in != nullptr;
    }

Reply:
Got it, thanks for clarification.